File: web/lib/shortcode.ts
// URL-safe short codes, mirroring the desktop app's qrgen/shortcode.py:
// 27-char alphabet with ambiguous chars (0 o 1 l i) removed.
const ALPHABET = "abcdefghjkmnpqrstuvwxyz23456789";
const DEFAULT_LENGTH = 7;

export function newShortCode(length = DEFAULT_LENGTH): string {
  const bytes = new Uint8Array(length);
  crypto.getRandomValues(bytes);
  let out = "";
  for (let i = 0; i < length; i++) {
    out += ALPHABET[bytes[i] % ALPHABET.length];
  }
  return out;
}

File: desktop-legacy/qrgen/shortcode.py
"""Short-code generation for dynamic QR codes.

Codes are URL-safe, lowercase, and avoid ambiguous characters (0/o/1/l/i) so they
stay readable if anyone ever has to type one. 7 chars over a 27-char alphabet is
~10 billion combinations — collisions are vanishingly unlikely but we still check
against the DB at creation time.
"""
from __future__ import annotations

import secrets

ALPHABET = "abcdefghjkmnpqrstuvwxyz23456789"  # no 0 o 1 l i
DEFAULT_LENGTH = 7


def new_code(length: int = DEFAULT_LENGTH) -> str:
    return "".join(secrets.choice(ALPHABET) for _ in range(length))
